packages/iframe-stamper/src/index.ts

```diff
@@ -0,0 +1,201 @@
+/// <reference lib="dom" />
+
+// Header name for an API key stamp
+const stampHeaderName = "X-Stamp";
+
+// Set of constants for event types expected to be sent and received between a parent page and its iframe.
+export enum IframeEventType {
+  // Event sent by the iframe to its parent to indicate readiness.
+  // Value: the iframe public key
+  PublicKeyReady = "PUBLIC_KEY_READY",
+  // Event sent by the parent to inject a recovery bundle into the iframe.
+  // Value: the bundle to inject
+  InjectRecoveryBundle = "INJECT_RECOVERY_BUNDLE",
+  // Event sent by the iframe to its parent when `InjectBundle` is successful
+  // Value: true (boolean)
+  BundleInjected = "BUNDLE_INJECTED",
+  // Event sent by the parent page to request a signature
+  // Value: payload to sign
+  StampRequest = "STAMP_REQUEST",
+  // Event sent by the iframe to communicate the result of a stamp operation.
+  // Value: signed payload
+  Stamp = "STAMP",
+}
+
+type TStamp = {
+  stampHeaderName: string;
+  stampHeaderValue: string;
+};
+
+export type TIframeStamperConfig = {
+  iframeUrl: string;
+  iframeElementId: string;
+  iframeContainerId: string;
+};
+
+/**
+ * Stamper to use with `@turnkey/http`'s `TurnkeyClient`
+ * Creating a stamper inserts an iframe in the current page.
+ */
+export class IframeStamper {
+  container: HTMLElement;
+  iframe: HTMLIFrameElement;
+  iframeOrigin: string;
+  iframePublicKey: string | null;
+
+  /**
+   * Creates a new iframe stamper. This function _does not_ insert the iframe in the DOM.
+   * Call `.init()` to insert the iframe element in the DOM.
+   */
+  constructor(config: TIframeStamperConfig) {
+    if (typeof window === "undefined") {
+      throw new Error("Cannot initialize iframe in non-browser environment");
+    }
+
+    if (document.getElementById(config.iframeElementId)) {
+      throw new Error(
+        `Iframe element with ID ${config.iframeElementId} already exists`
+      );
+    }
+
+    const container = document.getElementById(config.iframeContainerId);
+    if (!container) {
+      throw new Error(
+        `Cannot create iframe stamper: no container with ID ${config.iframeContainerId} exists in the current document`
+      );
+    }
+    this.container = container;
+
+    var iframe = window.document.createElement("iframe");
+    iframe.id = config.iframeElementId;
+    iframe.src = config.iframeUrl;
+    this.iframe = iframe;
+
+    const iframeUrl = new URL(config.iframeUrl);
+    this.iframeOrigin = iframeUrl.origin;
+
+    // This is populated once the iframe is ready. Call `.init()` to kick off DOM insertion!
+    this.iframePublicKey = null;
+  }
+
+  /**
+   * Inserts the iframe on the page and returns a promise resolving to the iframe's public key
+   */
+  async init(): Promise<string> {
+    this.container.appendChild(this.iframe);
```

Comment:
should we check that `this.container` is non null?

Reply:
yep yep! Good catch.

Reply:
Oh actually nevermind. This is already enforced by the non-null type. We already check for this in the constructor:
```ts
    const container = document.getElementById(config.iframeContainerId);
    if (!container) {
      throw new Error(
        `Cannot create iframe stamper: no container with ID ${config.iframeContainerId} exists in the current document`
      );
    }
    this.container = container;
```